master/src/main/scala/org/apache/celeborn/service/deploy/master/quota/QuotaManager.scala

```diff
@@ -16,96 +16,338 @@
   }
 
-  private def checkDiskBytesWritten(
-      userIdentifier: UserIdentifier,
+  private def checkQuota(
       value: Long,
-      quota: Quota): (Boolean, String) = {
-    val exceed = (quota.diskBytesWritten > 0 && value >= quota.diskBytesWritten)
+      quota: Long,
+      quotaType: String,
+      format: Long => String): (Boolean, String) = {
+    val exceed = quota > 0 && value >= quota
     var reason = ""
     if (exceed) {
-      reason = s"User $userIdentifier used diskBytesWritten (${Utils.bytesToString(value)}) " +
-        s"exceeds quota (${Utils.bytesToString(quota.diskBytesWritten)}). "
+      reason = s"$quotaType(${format(value)}) exceeds quota(${format(quota)}). "
       logWarning(reason)
     }
     (exceed, reason)
   }
 
-  private def checkDiskFileCount(
-      userIdentifier: UserIdentifier,
-      value: Long,
-      quota: Quota): (Boolean, String) = {
-    val exceed = (quota.diskFileCount > 0 && value >= quota.diskFileCount)
-    var reason = ""
-    if (exceed) {
-      reason =
-        s"User $userIdentifier used diskFileCount($value) exceeds quota(${quota.diskFileCount}). "
-      logWarning(reason)
+  private def checkConsumptionExceeded(
+      used: ResourceConsumption,
+      threshold: StorageQuota): Boolean = {
+    used.diskBytesWritten >= threshold.diskBytesWritten ||
+    used.diskFileCount >= threshold.diskFileCount ||
+    used.hdfsBytesWritten >= threshold.hdfsBytesWritten ||
+    used.hdfsFileCount >= threshold.hdfsFileCount
+  }
+
+  def updateResourceConsumption(): Unit = {
+    masterSource.sample(UPDATE_RESOURCE_CONSUMPTION_TIME, this.getClass.getSimpleName, Map.empty) {
+      val clusterQuota = getClusterStorageQuota
+      var clusterResourceConsumption = ResourceConsumption(0, 0, 0, 0)
+      val activeUsers = mutable.Set[UserIdentifier]()
+
+      val tenantResourceConsumption =
+        statusSystem.availableWorkers.asScala.flatMap { workerInfo =>
+          workerInfo.userResourceConsumption.asScala
+        }.groupBy(_._1.tenantId).toSeq.map { case (tenantId, tenantConsumptionList) =>
+          var tenantResourceConsumption = ResourceConsumption(0, 0, 0, 0)
+          val userResourceConsumption =
+            tenantConsumptionList.groupBy(_._1).map {
+              case (userIdentifier, userConsumptionList) =>
+                activeUsers.add(userIdentifier)
+                // Step 1: Compute user consumption and set quota status.
+                val resourceConsumptionList = userConsumptionList.map(_._2).toSeq
+                val resourceConsumption = computeUserResourceConsumption(resourceConsumptionList)
+
+                // Step 2: Update user resource consumption metrics.
+                // For extract metrics
+                userResourceConsumptionMap.put(userIdentifier, resourceConsumption)
+                registerUserResourceConsumptionMetrics(userIdentifier)
+
+                // Step 3: Expire user level exceeded app except already expired app
```

Comment:
Can we separate this into `checkUserResourceConsumption`, `checkTenantResourceConsumption`, and `checkClusterResourceConsumption`? This would improve unit testability and allow for better future extensions